cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java

```diff
@@ -0,0 +1,546 @@
+                                                                  RebalanceAnnotation rebalanceAnnotation) {
+        Promise<KafkaClusterRebalanceStatus> p = Promise.promise();
+        if (rebalanceAnnotation == RebalanceAnnotation.none) {
+            log.debug("Arming Cruise Control rebalance user task status timer");
+            String sessionId = clusterRebalance.getStatus().getSessionId();
+            vertx.setPeriodic(REBALANCE_TASK_STATUS_TIMER_MS, t -> {
+                resourceOperator.getAsync(clusterRebalance.getMetadata().getNamespace(), clusterRebalance.getMetadata().getName()).setHandler(getResult -> {
+                    if (getResult.succeeded()) {
+                        KafkaClusterRebalance fetchedClusterRebalance = getResult.result();
+                        if (getRebalanceAnnotation(fetchedClusterRebalance) == RebalanceAnnotation.stop) {
+                            log.debug("Stopping current Cruise Control rebalance user task");
+                            vertx.cancelTimer(t);
+                            apiClient.stopExecution("host", 0).setHandler(stopResult -> {
+                                p.complete(new KafkaClusterRebalanceStatusBuilder()
+                                        .withSessionId(null)
+                                        .addNewCondition().withNewType(State.Stopped.toString()).endCondition().build());
+                            });
+                        } else {
+                            log.debug("Getting Cruise Control rebalance user task status");
+                            apiClient.getUserTaskStatus("host", 0, sessionId).setHandler(userTaskResult -> {
+                                if (userTaskResult.succeeded()) {
+                                    CruiseControlResponse response = userTaskResult.result();
+                                    JsonObject taskStatusJson = response.getJson();
+                                    String taskStatus = taskStatusJson.getString("Status");
+                                    if (taskStatus.equals("completed")) {
+                                        vertx.cancelTimer(t);
+                                        p.complete(new KafkaClusterRebalanceStatusBuilder()
+                                                .withSessionId(null)
+                                                .withOptimizationResult(taskStatusJson.getJsonObject("rebalance").getMap())
+                                                .addNewCondition().withType(State.ProposalReady.toString()).endCondition().build());
+                                    } else if (taskStatus.equals("aborted")) {
+                                        log.error("Unexpected state {}", taskStatus);
+                                        vertx.cancelTimer(t);
+                                        p.fail("Unexpected state " + taskStatus);
+                                    }
```

Comment:
What other values can `taskStatus` have? If you are assuming that while the rebalance is ongoing that it will be `null` then any other non-null value should be treated as an error.

Reply:
from the CC code I see "aborting", "inprogress", "pending"

Reply:
So treat anything else (and `taskStatus == null`) as an error. Otherwise the periodic timer could run forever and the user will never know why it's not completing (and the only way to fix it is to restart the CO).

Reply:
well we need to do real tests with CC but "aborting", "inprogress", "pending" should still need the timer until it's "aborted" or "completed" (which I am handling right now).
Agree to refactor the code to handle `null`.

Reply:
My point is only that we should handle all the possible cases which we know about and error for anything else. We should not have any dangling case which results in what is, in effect, an infinite loop.